cps/services/worker.py

```diff
@@ -0,0 +1,184 @@
+
+QueuedTask = namedtuple('QueuedTask', 'user, added, task')
+
+
+def _get_main_thread():
+    for t in threading.enumerate():
+        if t.__class__.__name__ == '_MainThread':
+            return t
+    raise Exception("main thread not found?!")
+
+
+
+class ImprovedQueue(queue.Queue):
+    def to_list(self):
+        """
+        Returns a copy of all items in the queue without removing them.
+        """
+
+        with self.mutex:
+            return list(self.queue)
+
+#Class for all worker tasks in the background
+class WorkerThread(threading.Thread):
+    _instance = None
+
+    @classmethod
+    def getInstance(cls):
+        if cls._instance is None:
+            cls._instance = WorkerThread()
+        return cls._instance
+
+    def __init__(self):
+        threading.Thread.__init__(self)
+
+        self.dequeued = list()
+
+        self.doLock = threading.Lock()
+        self.queue = ImprovedQueue()
+
+        self.start()
+
+    @classmethod
+    def add(cls, user, task):
+        ins = cls.getInstance()
+        ins.queue.put(QueuedTask(
+            user=user,
+            added=datetime.now(),
+            task=task,
+        ))
+
+    @property
+    def tasks(self):
+        with self.doLock:
+            tasks = list(self.queue.to_list()) + self.dequeued
+        return sorted(tasks, key=lambda x: x.added)
+
+    # Main thread loop starting the different tasks
+    def run(self):
+        main_thread = _get_main_thread()
+        while main_thread.is_alive():
+            item = self.queue.get()
+
+            # add to list so that in-progress tasks show up
+            with self.doLock:
+                # Remove completed tasks if needed
+                if len(self.dequeued) > TASK_CLEANUP_TRIGGER:
+                    # sort first (just to be certain), then lob off the extra
+                    self.dequeued = sorted(self.dequeued, key=lambda x: x.added)[-1 * TASK_CLEANUP_TRIGGER:]
+                self.dequeued.append(item)
+
+            user, added, task = item
+
+            # sometimes tasks (like Upload) don't actually have work to do and are created as already finished
+            if task.stat is STAT_WAITING:
+                # CalibreTask.start() should wrap all exceptions in it's own error handling
+                task.start(self)
+
+            self.queue.task_done()
+
+class CalibreTask:
+    __metaclass__ = abc.ABCMeta
+
+    def __init__(self, message):
+        self._progress = 0
+        self.stat = STAT_WAITING
+        self.error = None
+        self.start_time = None
+        self.end_time = None
+        self.message = message
+        self.id = uuid.uuid4()
+
+    @abc.abstractmethod
+    def run(self, worker_thread):
+        """Provides the caller some human-readable name for this class"""
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def name(self):
+        """Provides the caller some human-readable name for this class"""
+        raise NotImplementedError
```

Comment:
These abstract methods should be overridden by the concrete class. All other things here in this class simply provide default / boilerplate code for each tasks (but they can also be overridden if needed, see: `TaskEmail.progress`)